storage/postgres/query_builder.go

```diff
@@ -0,0 +1,330 @@
+package postgres
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/Peripli/service-manager/pkg/query"
+	"github.com/Peripli/service-manager/pkg/util"
+	"github.com/jmoiron/sqlx"
+)
+
+type orderRule struct {
+	field     string
+	orderType query.OrderType
+}
+
+type QueryBuilder struct {
+	db pgDB
+
+	sql         string
+	queryParams []interface{}
+
+	labelCriteria, fieldCriteria []query.Criterion
+	orderByFields                []orderRule
+	limit                        int
+	criteria                     []query.Criterion
+	hasLock                      bool
+	returningFields              []string
+
+	err error
+}
+
+func NewQueryBuilder(db pgDB) *QueryBuilder {
+	return &QueryBuilder{
+		db: db,
+	}
+}
+
+func (qb *QueryBuilder) List(ctx context.Context, entity PostgresEntity) (*sqlx.Rows, error) {
+	if qb.err != nil {
```

Comment:
write some godoc somewhere that these methods are not thread-safe and the QueryBuilder cannot be reused.
I'd rather have queryBuilder in postgresql storage and reuse it instead of create new one every time.
The chaining can be extracted in another structure created once-per-build and `queryBuilder.NewQuery().WithCriteria()....` instead of `queryBuilder.WithCriteria()...`